From: Chad Monroe <chad.monroe@smartrg.com>
Date: Wed, 08 Sep 2021 12:50:14 -0700
Subject: btmtkuart: add set_bdaddr support so MAC can be changed 

Signed-off-by: Chad Monroe <chad.monroe@smartrg.com>
--
 btmtkuart.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/drivers/bluetooth/btmtkuart.c
+++ b/drivers/bluetooth/btmtkuart.c
@@ -44,6 +44,9 @@
 
 #define BTMTKUART_FLAG_STANDALONE_HW	 BIT(0)
 
+/* Vendor-specific HCI commands */
+#define HCI_VS_WRITE_BD_ADDR	0xfc1a
+
 enum {
 	MTK_WMT_PATCH_DWNLD = 0x1,
 	MTK_WMT_TEST = 0x2,
@@ -839,6 +842,20 @@ ignore_func_on:
 	return 0;
 }
 
+static int btmtkuart_set_bdaddr(struct hci_dev *hdev, const bdaddr_t *bdaddr)
+{
+	struct sk_buff *skb;
+
+	skb = __hci_cmd_sync(hdev, HCI_VS_WRITE_BD_ADDR, sizeof(bdaddr), bdaddr,
+						HCI_INIT_TIMEOUT);
+
+	if (IS_ERR(skb))
+		return PTR_ERR(skb);
+	kfree_skb(skb);
+
+	return 0;
+}
+
 static int btmtkuart_shutdown(struct hci_dev *hdev)
 {
 	struct btmtk_hci_wmt_params wmt_params;
@@ -1005,6 +1022,7 @@ static int btmtkuart_probe(struct serdev
 	hdev->close    = btmtkuart_close;
 	hdev->flush    = btmtkuart_flush;
 	hdev->setup    = btmtkuart_setup;
+	hdev->set_bdaddr = btmtkuart_set_bdaddr;
 	hdev->shutdown = btmtkuart_shutdown;
 	hdev->send     = btmtkuart_send_frame;
 	SET_HCIDEV_DEV(hdev, &serdev->dev);
