From 05f482dfab1f1437767e74ac1592c0cb917f57bd Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Tue, 28 Feb 2023 13:11:22 -0800
Subject: [PATCH] OWRT-7643 Adtran quirks added

---
 drivers/net/phy/sfp.c | 34 ++++++++++++++++++++++++++++++++--
 1 file changed, 32 insertions(+), 2 deletions(-)

--- a/drivers/net/phy/sfp.c
+++ b/drivers/net/phy/sfp.c
@@ -333,6 +333,7 @@ static void sfp_fixup_halny_gsfp(struct
 static void sfp_quirk_2500basex(const struct sfp_eeprom_id *id,
 				unsigned long *modes)
 {
+	linkmode_zero(modes);
 	linkmode_set_bit(ETHTOOL_LINK_MODE_2500baseX_Full_BIT, modes);
 }
 
@@ -347,20 +348,31 @@ static void sfp_quirk_ubnt_uf_instant(co
 	linkmode_set_bit(ETHTOOL_LINK_MODE_1000baseX_Full_BIT, modes);
 }
 
+static void sfp_fixup_gpon_xxx(struct sfp *sfp)
+{
+	sfp->module_t_start_up = T_START_UP_BAD_GPON;
+	sfp->tx_fault_ignore = true;
+	sfp->state_hw_mask &= ~(SFP_F_TX_FAULT | SFP_F_LOS);
+	printk("%s SFP \n",__FUNCTION__);
+}
+
 static const struct sfp_quirk sfp_quirks[] = {
 	{
 		// Alcatel Lucent G-010S-P can operate at 2500base-X, but
 		// incorrectly report 2500MBd NRZ in their EEPROM
+		// AKA Nokia from James Mouffat
 		.vendor = "ALCATELLUCENT",
 		.part = "G010SP",
 		.modes = sfp_quirk_2500basex,
+		.fixup = sfp_fixup_gpon_xxx,
 	}, {
 		// Alcatel Lucent G-010S-A can operate at 2500base-X, but
 		// report 3.2GBd NRZ in their EEPROM
+		// From Chad for reference
 		.vendor = "ALCATELLUCENT",
 		.part = "3FE46541AA",
 		.modes = sfp_quirk_2500basex,
-		.fixup = sfp_fixup_long_startup,
+		.fixup = sfp_fixup_gpon_xxx,
 	}, {
 		.vendor = "HALNy",
 		.part = "HL-GSFP",
@@ -371,7 +383,7 @@ static const struct sfp_quirk sfp_quirks
 		.vendor = "HUAWEI",
 		.part = "MA5671A",
 		.modes = sfp_quirk_2500basex,
-		.fixup = sfp_fixup_ignore_tx_fault,
+		.fixup = sfp_fixup_gpon_xxx,
 	}, {
 		// OEM SFP-GE-T is 1000Base-T module
 		.vendor = "OEM",
@@ -387,7 +399,25 @@ static const struct sfp_quirk sfp_quirks
 		.vendor = "UBNT",
 		.part = "UF-INSTANT",
 		.modes = sfp_quirk_ubnt_uf_instant,
+	}, {
+		// F2
+		.vendor = "T&W",
+		.part = "TW2362H-CDEL-ADT",
+		.modes = sfp_quirk_2500basex,
+		.fixup = sfp_fixup_gpon_xxx,
+	}, {
+		// F3
+		.vendor = "T&W",
+		.part = "TW2362H-CDEL-AD1",
+		.fixup = sfp_fixup_gpon_xxx,
+	}, {
+		// 630
+		.vendor = "ADTRAN",
+		.part = "1287940F1",
+		.modes = sfp_quirk_2500basex,
+		.fixup = sfp_fixup_gpon_xxx,
 	}
+
 };
 
 static size_t sfp_strlen(const char *str, size_t maxlen)
From c1a5765e7a9713b70fb4946dd296230024a03458 Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Thu, 1 Jun 2023 07:27:27 -0700
Subject: [PATCH] OWRT-8395 fixups for LINKTEL and FINISAR Copper SFPs

---
 drivers/net/phy/sfp.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/net/phy/sfp.c b/drivers/net/phy/sfp.c
index 42ad9c2..0125988 100644
--- a/drivers/net/phy/sfp.c
+++ b/drivers/net/phy/sfp.c
@@ -341,6 +341,15 @@ static void sfp_quirk_2500basex(const struct sfp_eeprom_id *id,
 {
 	linkmode_zero(modes);
 	linkmode_set_bit(ETHTOOL_LINK_MODE_2500baseX_Full_BIT, modes);
+	printk("%s SFP \n",__FUNCTION__);
+}
+
+static void sfp_quirk_1000basex(const struct sfp_eeprom_id *id,
+				unsigned long *modes)
+{
+	linkmode_zero(modes);
+	linkmode_set_bit(ETHTOOL_LINK_MODE_1000baseX_Full_BIT, modes);
+	printk("%s SFP \n",__FUNCTION__);
 }
 
 static void sfp_quirk_ubnt_uf_instant(const struct sfp_eeprom_id *id,
@@ -362,6 +371,12 @@ static void sfp_fixup_gpon_xxx(struct sfp *sfp)
 	printk("%s SFP \n",__FUNCTION__);
 }
 
+static void sfp_fixup_copper_xxx(struct sfp *sfp)
+{
+	sfp->id.base.e1000_base_t=0;
+	printk("%s SFP \n",__FUNCTION__);
+}
+
 static const struct sfp_quirk sfp_quirks[] = {
 	{
 		// Alcatel Lucent G-010S-P can operate at 2500base-X, but
@@ -426,6 +441,18 @@ static const struct sfp_quirk sfp_quirks[] = {
 		.part = "1287940F1",
 		.modes = sfp_quirk_2500basex,
 		.fixup = sfp_fixup_gpon_xxx,
+	}, {
+		// ADTRAN-1442300G1
+		.vendor = "LINKTEL",
+		.part = "LX1801INA-ADT",
+		.modes = sfp_quirk_1000basex,
+		.fixup = sfp_fixup_copper_xxx,
+	}, {
+		// FCLFB8521P2BTL-AR
+		.vendor = "FINISAR CORP.",
+		.part = "FCLF8521P2BTL-AR",
+		.modes = sfp_quirk_1000basex,
+		.fixup = sfp_fixup_copper_xxx,
 	}
 
 };
-- 
2.30.2

