Index: linux-5.15.72/drivers/net/phy/intel-xway.c
===================================================================
--- linux-5.15.72.orig/drivers/net/phy/intel-xway.c
+++ linux-5.15.72/drivers/net/phy/intel-xway.c
@@ -232,8 +232,6 @@ static int xway_gphy_rgmii_init(struct p
 static int xway_gphy_config_init(struct phy_device *phydev)
 {
 	int err;
-	u32 ledxh;
-	u32 ledxl;
 
 	/* Mask all interrupts */
 	err = phy_write(phydev, XWAY_MDIO_IMASK, 0);
@@ -266,15 +264,20 @@ static int xway_gphy_config_init(struct
 	 * only available in some packages, leave it in its reset
 	 * configuration.
 	 */
-	ledxh = XWAY_MMD_LEDxH_BLINKF_NONE | XWAY_MMD_LEDxH_CON_LINK10XX;
-	ledxl = XWAY_MMD_LEDxL_PULSE_TXACT | XWAY_MMD_LEDxL_PULSE_RXACT |
-		XWAY_MMD_LEDxL_BLINKS_NONE;
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED0H, ledxh);
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED0L, ledxl);
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED1H, ledxh);
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED1L, ledxl);
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED2H, ledxh);
-	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED2L, ledxl);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED0H,
+			XWAY_MMD_LEDxH_BLINKF_NONE | XWAY_MMD_LEDxH_CON_LINK1000);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED0L,
+			XWAY_MMD_LEDxL_BLINKS_NONE);
+
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED1H,
+			XWAY_MMD_LEDxH_BLINKF_NONE | XWAY_MMD_LEDxH_CON_LINK10X);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED1L,
+			XWAY_MMD_LEDxL_BLINKS_NONE);
+
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED2H,
+			XWAY_MMD_LEDxH_BLINKF_NONE | XWAY_MMD_LEDxH_CON_NONE);
+	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LED2L,
+			XWAY_MMD_LEDxL_BLINKS_NONE);
 
 	err = xway_gphy_rgmii_init(phydev);
 	if (err)
