--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -3283,6 +3283,9 @@ static void mtk_tx_timeout(struct net_de
 	eth->netdev[mac->id]->stats.tx_errors++;
 	netif_err(eth, tx_err, dev, "transmit timed out\n");
 
+	dev_err(eth->dev, "transmit queue timed out.. PANIC!\n");
+	panic("mtk_soc_eth transmit queue timed out");
+
 	schedule_work(&eth->pending_work);
 }
 
--- a/drivers/net/ethernet/mediatek/mtk_ppe.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.c
@@ -69,8 +69,10 @@ static int mtk_ppe_wait_busy(struct mtk_
 				 !(val & MTK_PPE_GLO_CFG_BUSY),
 				 20, MTK_PPE_WAIT_TIMEOUT_US);
 
-	if (ret)
-		dev_err(ppe->dev, "PPE table busy");
+	if (ret) {
+		dev_err(ppe->eth->dev, "PPE table busy.. PANIC!\n");
+		panic("mtk_soc_eth PPE table busy");
+	}
 
 	return ret;
 }
