--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -3283,6 +3283,9 @@ static void mtk_tx_timeout(struct net_de
 	eth->netdev[mac->id]->stats.tx_errors++;
 	netif_err(eth, tx_err, dev, "transmit timed out\n");
 
+	dev_err(eth->dev, "transmit queue timed out.. PANIC!\n");
+	panic("mtk_soc_eth transmit queue timed out");
+
 	schedule_work(&eth->pending_work);
 }
 
