--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -884,7 +884,28 @@ static int aqr113c_fill_interface_modes(
 	if (ret)
 		return ret;
 
-	return aqr107_fill_interface_modes(phydev);
+	ret = aqr107_fill_interface_modes(phydev);
+	if (ret)
+		return ret;
+	return aqr_hwmon_probe(phydev);
+}
+
+static int aqr113c_probe(struct phy_device *phydev)
+{
+	int ret;
+
+	phydev->priv = devm_kzalloc(&phydev->mdio.dev,
+				    sizeof(struct aqr107_priv), GFP_KERNEL);
+	if (!phydev->priv)
+		return -ENOMEM;
+
+	ret = aqr_firmware_load(phydev);
+	if (ret)
+		return ret;
+
+	aqr_mib_probe(phydev);
+
+	return 0;
 }
 
 static int aqr113c_config_init(struct phy_device *phydev)
@@ -1120,7 +1141,7 @@ static struct phy_driver aqr_driver[] =
 {
 	PHY_ID_MATCH_MODEL(PHY_ID_AQR113),
 	.name		= "Aquantia AQR113",
-	.probe          = aqr107_probe,
+	.probe          = aqr113c_probe,
 	.remove		= aqr107_remove,
 	.get_rate_matching = aqr107_get_rate_matching,
 	.config_init    = aqr113c_config_init,
@@ -1145,7 +1166,7 @@ static struct phy_driver aqr_driver[] =
 {
 	PHY_ID_MATCH_MODEL(PHY_ID_AQR113C),
 	.name           = "Aquantia AQR113C",
-	.probe          = aqr107_probe,
+	.probe          = aqr113c_probe,
 	.remove			= aqr107_remove,
 	.get_rate_matching = aqr107_get_rate_matching,
 	.config_init    = aqr113c_config_init,
