--- a/drivers/pinctrl/mediatek/pinctrl-mt7987.c
+++ b/drivers/pinctrl/mediatek/pinctrl-mt7987.c
@@ -476,12 +476,13 @@ static int mt7987_watchdog_funcs[] = {1}
 static int mt7987_jtag_pins[] = {3, 4, 5, 6, 7};
 static int mt7987_jtag_funcs[] = {1, 1, 1, 1, 1};
 
-/* pcm */
-static int mt7987_pcm0_0_pins[] = {3, 4, 5, 6, 7};
-static int mt7987_pcm0_0_funcs[] = {2, 2, 2, 2, 2};
+/* i2s */
+static int mt7987_i2s_pins[] = { 8, 9, 10, 11, 12 };
+static int mt7987_i2s_funcs[] = { 1, 1, 1, 1, 1 };
 
-static int mt7987_pcm0_1_pins[] = {8, 9, 10, 11, 12};
-static int mt7987_pcm0_1_funcs[] = {1, 1, 1, 1, 1};
+/* pcm */
+static int mt7987_pcm_pins[] = { 8, 9, 10, 11 };
+static int mt7987_pcm_funcs[] = { 1, 1, 1, 1 };
 
 /* uart */
 static int mt7987_uart0_pins[] = {31, 32};
@@ -607,8 +608,8 @@ static int mt7987_pcie_phy_i2c_funcs[] =
 static const struct group_desc mt7987_groups[] = {
 	PINCTRL_PIN_GROUP("watchdog", mt7987_watchdog),
 	PINCTRL_PIN_GROUP("jtag", mt7987_jtag),
-	PINCTRL_PIN_GROUP("pcm0_0", mt7987_pcm0_0),
-	PINCTRL_PIN_GROUP("pcm0_1", mt7987_pcm0_1),
+	PINCTRL_PIN_GROUP("i2s", mt7987_i2s),
+	PINCTRL_PIN_GROUP("pcm", mt7987_pcm),
 	PINCTRL_PIN_GROUP("uart0", mt7987_uart0),
 	PINCTRL_PIN_GROUP("uart1_0", mt7987_uart1_0),
 	PINCTRL_PIN_GROUP("uart1_0_lite", mt7987_uart1_0_lite),
@@ -650,7 +651,6 @@ static const struct group_desc mt7987_gr
 
 static const char *const mt7987_wdt_groups[] = {"watchdog",};
 static const char *const mt7987_jtag_groups[] = {"jtag",};
-static const char *const mt7987_pcm_groups[] = {"pcm0_0", "pcm0_1"};
 static const char *const mt7987_uart_groups[] = {"uart0", "uart1_0",
 						 "uart1_0_lite", "uart1_1",
 						 "uart1_2", "uart2_0",
@@ -669,13 +669,15 @@ static const char *const mt7987_pcie_gro
 						 "pcie0_clkreq", "pcie0_wake",
 						 "pcie1_pereset", "pcie1_clkreq",
 						 "pcie1_wake",};
-static const char *const mt7987_i2s_groups[] = {"pcm0_0", "pcm0_1"};
 
+static const char * const mt7987_audio_groups[] = {
+	"i2s", "pcm",
+};
 
 static const struct function_desc mt7987_functions[] = {
+	{"audio", mt7987_audio_groups, ARRAY_SIZE(mt7987_audio_groups)},
 	{"wdt", mt7987_wdt_groups, ARRAY_SIZE(mt7987_wdt_groups)},
 	{"jtag", mt7987_jtag_groups, ARRAY_SIZE(mt7987_jtag_groups)},
-	{"pcm", mt7987_pcm_groups, ARRAY_SIZE(mt7987_pcm_groups)},
 	{"uart", mt7987_uart_groups, ARRAY_SIZE(mt7987_uart_groups)},
 	{"pwm", mt7987_pwm_groups, ARRAY_SIZE(mt7987_pwm_groups)},
 	{"usb", mt7987_usb_groups, ARRAY_SIZE(mt7987_usb_groups)},
@@ -685,7 +687,6 @@ static const struct function_desc mt7987
 	{"flash", mt7987_flash_groups, ARRAY_SIZE(mt7987_flash_groups)},
 	{"i2c", mt7987_i2c_groups, ARRAY_SIZE(mt7987_i2c_groups)},
 	{"pcie", mt7987_pcie_groups, ARRAY_SIZE(mt7987_pcie_groups)},
-	{"i2s", mt7987_i2s_groups, ARRAY_SIZE(mt7987_i2s_groups)},
 };
 
 static const struct mtk_eint_hw mt7987_eint_hw = {
