From 6772d4fcd78551de35872bf255ceb6cf5aeb4a40 Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Thu, 16 Dec 2021 13:02:28 -0800
Subject: [PATCH] OWRT-6281 added gpy_lpa_fix to gpa_probe

---
 drivers/net/phy/mxl-gpy.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/net/phy/mxl-gpy.c b/drivers/net/phy/mxl-gpy.c
index bfc7fc6..84ecd57 100644
--- a/drivers/net/phy/mxl-gpy.c
+++ b/drivers/net/phy/mxl-gpy.c
@@ -95,10 +95,25 @@ static int gpy_config_init(struct phy_device *phydev)
 	return ret < 0 ? ret : 0;
 }
 
+static void gpy_lpa_fix(struct phy_device *phydev)
+{
+	u32 val=0;
+	val = phy_read(phydev, MII_LPA);
+	if (val==0) {
+		val=0x11e0;
+		phy_write(phydev, MII_LPA,val);
+		val = phy_read(phydev, MII_LPA);
+		printk("%s:%d  LPA zero fix activated val 0x%x phydev->autoneg %d\n",__FUNCTION__,__LINE__,val,phydev->autoneg);
+	}
+}
+
+
 static int gpy_probe(struct phy_device *phydev)
 {
 	int ret;
 
+	gpy_lpa_fix(phydev);
+
 	if (!phydev->is_c45) {
 		ret = phy_get_c45_ids(phydev);
 		if (ret < 0)
-- 
2.20.1

