Index: linux-5.10.100/drivers/net/phy/phylink.c
===================================================================
--- linux-5.10.100.orig/drivers/net/phy/phylink.c
+++ linux-5.10.100/drivers/net/phy/phylink.c
@@ -1064,7 +1064,7 @@ static int phylink_bringup_phy(struct ph
 			     __ETHTOOL_LINK_MODE_MASK_NBITS, phy->supported,
 			     __ETHTOOL_LINK_MODE_MASK_NBITS, config.advertising,
 			     ret);
-		return ret;
+	//	return ret;
 	}
 
 	phy->phylink = pl;
