From f719858281c045cee2ef364761acd4a577adbef8 Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Fri, 28 Jul 2023 11:49:52 -0700
Subject: [PATCH] OWRT-8586 xcon wan override

---
 net/bridge/br_switchdev.c | 9 +++++++++
 1 file changed, 9 insertions(+)

--- a/net/bridge/br_switchdev.c
+++ b/net/bridge/br_switchdev.c
@@ -66,6 +66,15 @@ bool nbp_switchdev_allowed_egress(const
 {
 	struct br_input_skb_cb *cb = BR_INPUT_SKB_CB(skb);
 
+	if (strncmp(p->br->dev->name,"br-xcon",sizeof("br-xcon")) >=0 ) {
+		if (strncmp(p->dev->name,"wan",sizeof("wan")) >=0 ) {
+			return(true);
+		}
+		if (strncmp(skb->dev->name,"wan",sizeof("wan")) >=0 ) {
+			return(true);
+		}
+	}
+
 	return !test_bit(p->hwdom, &cb->fwd_hwdoms) &&
 		(!skb->offload_fwd_mark || cb->src_hwdom != p->hwdom);
 }
