--- a/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7622-rfb1.dts
@@ -59,6 +59,33 @@
 		reg = <0 0x40000000 0 0x40000000>;
 	};
 
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		/delete-node/ramoops@42ff0000;
+
+		bootdata_reserved: bootdata@0x45000000 {
+			no-map;
+			reg = <0x0 0x45000000 0x0 0x00001000>;
+		};
+		ramoops_reserved: ramoops1@0x45001000 {
+			no-map;
+			compatible = "ramoops";
+			reg = <0x0 0x45001000 0x0 0x00140000>;
+			ftrace-size = <0x20000>;
+			record-size = <0x20000>;
+			console-size = <0x20000>;
+			pmsg-size = <0x80000>;
+		};
+	};
+
+	bootdata {
+		compatible = "bootdata";
+		memory-region = <&bootdata_reserved>;
+	};
+
 	reg_1p8v: regulator-1p8v {
 		compatible = "regulator-fixed";
 		regulator-name = "fixed-1.8V";
@@ -164,6 +191,20 @@
 					label = "wan";
 				};
 
+				port@5 {
+					reg = <5>;
+					label = "lan5";
+					phy-mode = "2500base-x";
+
+					compatible = "maxlinear,gpy211", "ethernet-phy-ieee802.3-c22";
+
+					fixed-link {
+						speed = <2500>;
+						full-duplex;
+						pause;
+					};
+				};
+
 				port@6 {
 					reg = <6>;
 					label = "cpu";
@@ -259,6 +300,21 @@
 	status = "okay";
 };
 
+&slot0 {
+	mt76@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		/* mediatek,mtd-eeprom = <&factory 0x0000>; */
+	};
+};
+
+&slot1 {
+	mt76@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		/* mediatek,mtd-eeprom = <&factory 0x8000>; */
+		ieee80211-freq-limit = <5000000 6000000>;
+	};
+};
+
 &pio {
 	/* Attention: GPIO 90 is used to switch between PCIe@1,0 and
 	 * SATA functions. i.e. output-high: PCIe, output-low: SATA
@@ -601,12 +657,17 @@
 
 			partition@200000 {
 				label = "firmware";
-				reg = <0x2c0000 0x2000000>;
+				reg = <0x2c0000 0x4000000>;
 			};
 
 			partition@2200000 {
-				label = "User_data";
-				reg = <0x22c0000 0x4000000>;
+				label = "nand";
+				reg = <0x42c0000 0x3140000>;
+			};
+
+			partition@7400000 {
+				label = "mfginfo";
+				reg = <0x7400000 0x100000>;
 			};
 		};
 	};
@@ -661,5 +722,5 @@
 	pinctrl-names = "default";
 	pinctrl-0 = <&wmac_pins>;
 	mediatek,mtd-eeprom = <&factory 0x0000>;
-	status = "okay";
+	status = "disabled";
 };
