--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -29,7 +29,7 @@ static int __airoha_npu_send_msg(struct
 {
 	struct airoha_npu_core *core = &npu->cores[0]; /* FIXME: core */
 	u32 val, offset = 0;
-	int ret;
+	int ret, retry_cnt=0, is_err=0;
 
 	if (len > AIROHA_NPU_MBOX_SIZE)
 		return -EINVAL;
@@ -49,17 +49,40 @@ static int __airoha_npu_send_msg(struct
 
 	memcpy(core->buf, data, len);
 
+npu_send_msg_retry:
 	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(0) + offset, core->addr);
 	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(1) + offset, len);
-	regmap_read(npu->regmap, REG_CR_MBQ0_CTRL(2) + offset, &val);
-	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(2) + offset, val + 1);
 	val = FIELD_PREP(MBOX_MSG_FUNC_ID, func_id) | MBOX_MSG_WAIT_RSP;
 	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(3) + offset, val);
+	regmap_read(npu->regmap, REG_CR_MBQ0_CTRL(2) + offset, &val);
+	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(2) + offset, val + 1);
 
 	ret = regmap_read_poll_timeout_atomic(npu->regmap,
 					      REG_CR_MBQ0_CTRL(3) + offset,
 					      val, (val & MBOX_MSG_DONE),
-					      100, 500 * MSEC_PER_SEC);
+					      100, (1000 * MSEC_PER_SEC));
+    is_err=0;
+	if (!ret && FIELD_GET(MBOX_MSG_STATUS, val) != NPU_MBOX_SUCCESS) {
+		retry_cnt++;
+		is_err=1;
+	    printk("(%s) npu done but failed at task_exec! (ret:%d, retry:%d, val:0x%x)\n", __func__, ret, retry_cnt, val);
+	}
+	else if (ret) {
+	    retry_cnt++;
+	    is_err=1;
+	    printk("(%s) regmap_read fail or timeout due to npu not done! (ret:%d, retry:%d)\n", __func__, ret, retry_cnt);
+	}
+	else {}
+
+    if (is_err==0) {}
+    else {
+    	if (retry_cnt<=3) {
+            goto npu_send_msg_retry;
+    	}
+    	else {
+            ret = -EINVAL;
+    	}
+	}
 	arht_npu_retry_send_msg(npu, ret, val, core->addr, len, func_id);
 
 	/* Copy the trailing reply_len bytes of the response. */
@@ -468,11 +491,10 @@ static int airoha_npu_wlan_init_memory(s
 	if (of_property_match_string(npu->dev->of_node, "memory-region-names",
 				     "ba") >= 0) {
 		cmd = WLAN_FUNC_SET_WAIT_DRAM_BA_NODE_ADDR;
-		err = airoha_npu_wlan_set_reserved_memory(npu, 0, "ba", cmd);
+		err = airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "ba", 0, &val);
 		if (err)
 			return err;
 	}
-
 	cmd = WLAN_FUNC_SET_WAIT_IS_FORCE_TO_CPU;
 	return airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "force_to_cpu", 1, &val);
 }
