--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -25,7 +25,7 @@
 
 static int __airoha_npu_send_msg(struct airoha_npu *npu, int func_id,
 				 const void *data, int len, void *reply,
-				 u16 reply_len)
+				 u16 reply_len, bool need_retry)
 {
 	struct airoha_npu_core *core = &npu->cores[0]; /* FIXME: core */
 	u32 val, offset = 0;
@@ -47,9 +47,9 @@ static int __airoha_npu_send_msg(struct
 	 */
 	spin_lock_bh(&core->lock);
 
+npu_send_msg_retry:
 	memcpy(core->buf, data, len);
 
-npu_send_msg_retry:
 	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(0) + offset, core->addr);
 	regmap_write(npu->regmap, REG_CR_MBQ0_CTRL(1) + offset, len);
 	val = FIELD_PREP(MBOX_MSG_FUNC_ID, func_id) | MBOX_MSG_WAIT_RSP;
@@ -60,7 +60,12 @@ npu_send_msg_retry:
 	ret = regmap_read_poll_timeout_atomic(npu->regmap,
 					      REG_CR_MBQ0_CTRL(3) + offset,
 					      val, (val & MBOX_MSG_DONE),
-					      100, (1000 * MSEC_PER_SEC));
+					      100, (need_retry ? 1000 : 500) * MSEC_PER_SEC);
+	if (!need_retry) {
+		if (!ret && FIELD_GET(MBOX_MSG_STATUS, val) != NPU_MBOX_SUCCESS)
+			ret = -EINVAL;
+		goto npu_send_msg_done;
+	}
     is_err=0;
 	if (!ret && FIELD_GET(MBOX_MSG_STATUS, val) != NPU_MBOX_SUCCESS) {
 		retry_cnt++;
@@ -85,6 +90,7 @@ npu_send_msg_retry:
 	}
 	arht_npu_retry_send_msg(npu, ret, val, core->addr, len, func_id);
 
+npu_send_msg_done:
 	/* Copy the trailing reply_len bytes of the response. */
 	if (!ret && reply)
 		memcpy(reply, core->buf + len - reply_len, reply_len);
@@ -97,7 +103,13 @@ npu_send_msg_retry:
 int airoha_npu_send_msg(struct airoha_npu *npu, int func_id,
 			void *data, int len)
 {
-	return __airoha_npu_send_msg(npu, func_id, data, len, data, len);
+	return __airoha_npu_send_msg(npu, func_id, data, len, data, len, false);
+}
+
+int airoha_npu_send_msg_need_retry(struct airoha_npu *npu, int func_id,
+				   void *data, int len)
+{
+	return __airoha_npu_send_msg(npu, func_id, data, len, data, len, true);
 }
 
 static int airoha_npu_load_firmware(struct device *dev, void __iomem *addr,
@@ -379,7 +391,7 @@ int airoha_npu_wlan_msg_send(struct airo
 	wlan_data->func_id = func_id;
 	memcpy(wlan_data->d, data, data_len);
 
-	err = airoha_npu_send_msg(npu, NPU_FUNC_WIFI, wlan_data, len);
+	err = airoha_npu_send_msg_need_retry(npu, NPU_FUNC_WIFI, wlan_data, len);
 	kfree(wlan_data);
 
 	return err;
@@ -402,7 +414,7 @@ int airoha_npu_wlan_msg_get(struct airoh
 	wlan_data->func_id = func_id;
 
 	err = __airoha_npu_send_msg(npu, NPU_FUNC_WIFI, wlan_data, len,
-				    data, data_len);
+				    data, data_len, true);
 	kfree(wlan_data);
 
 	return err;
