--- a/net/bridge/br_device.c
+++ b/net/bridge/br_device.c
@@ -495,7 +495,7 @@ void br_dev_setup(struct net_device *dev
 	dev->features = COMMON_FEATURES | NETIF_F_HW_VLAN_CTAG_TX |
 			NETIF_F_HW_VLAN_STAG_TX;
 	dev->hw_features = COMMON_FEATURES | NETIF_F_HW_VLAN_CTAG_TX |
-			   NETIF_F_HW_VLAN_STAG_TX;
+			   NETIF_F_HW_VLAN_STAG_TX | NETIF_F_LRO;
 	dev->vlan_features = COMMON_FEATURES;
 	dev->pcpu_stat_type = NETDEV_PCPU_STAT_TSTATS;
 
