--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -13,154 +13,9 @@
 #include <linux/regmap.h>
 
 #include "airoha_eth.h"
+#include "airoha_npu.h"
 
-#define NPU_EN7581_FIRMWARE_DATA		"airoha/en7581_npu_data.bin"
-#define NPU_EN7581_FIRMWARE_RV32		"airoha/en7581_npu_rv32.bin"
-#define NPU_EN7581_7996_FIRMWARE_DATA		"airoha/en7581_MT7996_npu_data.bin"
-#define NPU_EN7581_7996_FIRMWARE_RV32		"airoha/en7581_MT7996_npu_rv32.bin"
-#define NPU_AN7583_FIRMWARE_DATA		"airoha/an7583_npu_data.bin"
-#define NPU_AN7583_FIRMWARE_RV32		"airoha/an7583_npu_rv32.bin"
-#define NPU_EN7581_FIRMWARE_RV32_MAX_SIZE	0x200000
-#define NPU_EN7581_FIRMWARE_DATA_MAX_SIZE	0x10000
-#define NPU_DUMP_SIZE				512
-
-#define REG_NPU_LOCAL_SRAM		0x0
-
-#define NPU_PC_BASE_ADDR		0x305000
-#define REG_PC_DBG(_n)			(0x305000 + ((_n) * 0x100))
-
-#define NPU_CLUSTER_BASE_ADDR		0x306000
-
-#define REG_CR_BOOT_TRIGGER		(NPU_CLUSTER_BASE_ADDR + 0x000)
-#define REG_CR_BOOT_CONFIG		(NPU_CLUSTER_BASE_ADDR + 0x004)
-#define REG_CR_BOOT_BASE(_n)		(NPU_CLUSTER_BASE_ADDR + 0x020 + ((_n) << 2))
-
-#define NPU_MBOX_BASE_ADDR		0x30c000
-
-#define REG_CR_MBOX_INT_STATUS		(NPU_MBOX_BASE_ADDR + 0x000)
-#define MBOX_INT_STATUS_MASK		BIT(8)
-
-#define REG_CR_MBOX_INT_MASK(_n)	(NPU_MBOX_BASE_ADDR + 0x004 + ((_n) << 2))
-#define REG_CR_MBQ0_CTRL(_n)		(NPU_MBOX_BASE_ADDR + 0x030 + ((_n) << 2))
-#define REG_CR_MBQ8_CTRL(_n)		(NPU_MBOX_BASE_ADDR + 0x0b0 + ((_n) << 2))
-#define REG_CR_NPU_MIB(_n)		(NPU_MBOX_BASE_ADDR + 0x140 + ((_n) << 2))
-
-#define NPU_WLAN_BASE_ADDR		0x30d000
-
-#define REG_IRQ_STATUS			(NPU_WLAN_BASE_ADDR + 0x030)
-#define REG_IRQ_RXDONE(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 2) + 0x034)
-#define NPU_IRQ_RX_MASK(_n)		((_n) == 1 ? BIT(17) : BIT(16))
-
-#define REG_TX_BASE(_n)			(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x080)
-#define REG_TX_DSCP_NUM(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x084)
-#define REG_TX_CPU_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x088)
-#define REG_TX_DMA_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x08c)
-
-#define REG_RX_BASE(_n)			(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x180)
-#define REG_RX_DSCP_NUM(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x184)
-#define REG_RX_CPU_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x188)
-#define REG_RX_DMA_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x18c)
-
-#define NPU_TIMER_BASE_ADDR		0x310100
-#define REG_WDT_TIMER_CTRL(_n)		(NPU_TIMER_BASE_ADDR + ((_n) * 0x100))
-#define WDT_EN_MASK			BIT(25)
-#define WDT_INTR_MASK			BIT(21)
-
-enum {
-	NPU_OP_SET = 1,
-	NPU_OP_SET_NO_WAIT,
-	NPU_OP_GET,
-	NPU_OP_GET_NO_WAIT,
-};
-
-enum {
-	NPU_FUNC_WIFI,
-	NPU_FUNC_TUNNEL,
-	NPU_FUNC_NOTIFY,
-	NPU_FUNC_DBA,
-	NPU_FUNC_TR471,
-	NPU_FUNC_PPE,
-};
-
-enum {
-	NPU_MBOX_ERROR,
-	NPU_MBOX_SUCCESS,
-};
-
-enum {
-	PPE_FUNC_SET_WAIT,
-	PPE_FUNC_SET_WAIT_HWNAT_INIT,
-	PPE_FUNC_SET_WAIT_HWNAT_DEINIT,
-	PPE_FUNC_SET_WAIT_API,
-	PPE_FUNC_SET_WAIT_FLOW_STATS_SETUP,
-};
-
-enum {
-	PPE2_SRAM_SET_ENTRY,
-	PPE_SRAM_SET_ENTRY,
-	PPE_SRAM_SET_VAL,
-	PPE_SRAM_RESET_VAL,
-};
-
-enum {
-	QDMA_WAN_ETHER = 1,
-	QDMA_WAN_PON_XDSL,
-};
-
-struct airoha_npu_fw {
-	const char *name;
-	int max_size;
-};
-
-struct airoha_npu_soc_data {
-	struct airoha_npu_fw fw_rv32;
-	struct airoha_npu_fw fw_data;
-};
-
-#define MBOX_MSG_FUNC_ID	GENMASK(14, 11)
-#define MBOX_MSG_STATIC_BUF	BIT(5)
-#define MBOX_MSG_STATUS		GENMASK(4, 2)
-#define MBOX_MSG_DONE		BIT(1)
-#define MBOX_MSG_WAIT_RSP	BIT(0)
-
-#define PPE_TYPE_L2B_IPV4	2
-#define PPE_TYPE_L2B_IPV4_IPV6	3
-
-struct ppe_mbox_data {
-	u32 func_type;
-	u32 func_id;
-	union {
-		struct {
-			u8 cds;
-			u8 xpon_hal_api;
-			u8 wan_xsi;
-			u8 ct_joyme4;
-			u8 max_packet;
-			u8 rsv[3];
-			u32 ppe_type;
-			u32 wan_mode;
-			u32 wan_sel;
-		} init_info;
-		struct {
-			u32 func_id;
-			u32 size;
-			u32 data;
-		} set_info;
-		struct {
-			u32 npu_stats_addr;
-			u32 foe_stats_addr;
-		} stats_info;
-	};
-};
-
-struct wlan_mbox_data {
-	u32 ifindex:4;
-	u32 func_type:4;
-	u32 func_id;
-	DECLARE_FLEX_ARRAY(u8, d);
-};
-
-static int airoha_npu_send_msg(struct airoha_npu *npu, int func_id,
+int airoha_npu_send_msg(struct airoha_npu *npu, int func_id,
 			       void *p, int size)
 {
 	u16 core = 0; /* FIXME */
@@ -185,9 +40,7 @@ static int airoha_npu_send_msg(struct ai
 	ret = regmap_read_poll_timeout_atomic(npu->regmap,
 					      REG_CR_MBQ0_CTRL(3) + offset,
 					      val, (val & MBOX_MSG_DONE),
-					      100, 100 * MSEC_PER_SEC);
-	if (!ret && FIELD_GET(MBOX_MSG_STATUS, val) != NPU_MBOX_SUCCESS)
-		ret = -EINVAL;
+					      100, 500 * MSEC_PER_SEC);
 
 	spin_unlock_bh(&npu->cores[core].lock);
 
@@ -458,7 +311,7 @@ out:
 	return err;
 }
 
-static int airoha_npu_wlan_msg_send(struct airoha_npu *npu, int ifindex,
+int airoha_npu_wlan_msg_send(struct airoha_npu *npu, int ifindex,
 				    enum airoha_npu_wlan_set_cmd func_id,
 				    void *data, int data_len, gfp_t gfp)
 {
@@ -481,7 +334,7 @@ static int airoha_npu_wlan_msg_send(stru
 	return err;
 }
 
-static int airoha_npu_wlan_msg_get(struct airoha_npu *npu, int ifindex,
+int airoha_npu_wlan_msg_get(struct airoha_npu *npu, int ifindex,
 				   enum airoha_npu_wlan_get_cmd func_id,
 				   void *data, int data_len, gfp_t gfp)
 {
@@ -525,29 +378,63 @@ airoha_npu_wlan_set_reserved_memory(stru
 					sizeof(val), GFP_KERNEL);
 }
 
+
+static int airoha_npu_wait_init_complete(struct airoha_npu *npu)
+{
+	int reg_val;
+
+	regmap_read(npu->regmap, CR_NPU_MIB31, &reg_val);
+	while (reg_val != INIT_COMPELET) {
+		msleep(100);
+		regmap_read(npu->regmap, CR_NPU_MIB31, &reg_val);
+	}
+	return 0;
+}
+static int airoha_npu_wlan_cmd_with_retry(struct airoha_npu *npu, int ifindex,
+					  enum airoha_npu_wlan_set_cmd cmd,
+					  const char *name, int is_msg_send, u32 *val)
+{
+	int err = 0, retry_cnt = 3;
+
+	while (retry_cnt) {
+		if (is_msg_send)
+			err = airoha_npu_wlan_msg_send(npu, ifindex, cmd, val, sizeof(*val), GFP_KERNEL);
+		else
+			err = airoha_npu_wlan_set_reserved_memory(npu, ifindex, name, cmd);
+		if (!err)
+			break;
+		mdelay(10);
+		retry_cnt--;
+	}
+	mdelay(10);
+	if (!retry_cnt)
+		return err;
+	return 0;
+}
+
 static int airoha_npu_wlan_init_memory(struct airoha_npu *npu)
 {
 	enum airoha_npu_wlan_set_cmd cmd = WLAN_FUNC_SET_WAIT_NPU_BAND0_ONCPU;
 	u32 val = 0;
-	int err;
+	int err = 0;
 
-	err = airoha_npu_wlan_msg_send(npu, 1, cmd, &val, sizeof(val),
-				       GFP_KERNEL);
+	airoha_npu_wait_init_complete(npu);
+	err = airoha_npu_wlan_cmd_with_retry(npu, 1, cmd, "band0_oncpu", 1, &val);
 	if (err)
 		return err;
 
 	cmd = WLAN_FUNC_SET_WAIT_TX_BUF_CHECK_ADDR;
-	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "tx-bufid", cmd);
+	err = airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "tx-bufid", 0, NULL);
 	if (err)
 		return err;
 
 	cmd = WLAN_FUNC_SET_WAIT_PKT_BUF_ADDR;
-	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "pkt", cmd);
+	err = airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "pkt", 0, NULL);
 	if (err)
 		return err;
 
 	cmd = WLAN_FUNC_SET_WAIT_TX_PKT_BUF_ADDR;
-	err = airoha_npu_wlan_set_reserved_memory(npu, 0, "tx-pkt", cmd);
+	err = airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "tx-pkt", 0, NULL);
 	if (err)
 		return err;
 
@@ -560,8 +447,7 @@ static int airoha_npu_wlan_init_memory(s
 	}
 
 	cmd = WLAN_FUNC_SET_WAIT_IS_FORCE_TO_CPU;
-	return airoha_npu_wlan_msg_send(npu, 0, cmd, &val, sizeof(val),
-					GFP_KERNEL);
+	return airoha_npu_wlan_cmd_with_retry(npu, 0, cmd, "force_to_cpu", 1, &val);
 }
 
 static u32 airoha_npu_wlan_queue_addr_get(struct airoha_npu *npu, int qid,
--- /dev/null
+++ b/drivers/net/ethernet/airoha/airoha_npu.h
@@ -0,0 +1,186 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * Copyright (c) 2024 AIROHA Inc
+ */
+#ifndef AIROHA_NPU_H
+#define AIROHA_NPU_H
+
+#define NPU_EN7581_FIRMWARE_DATA		"airoha/en7581_npu_data.bin"
+#define NPU_EN7581_FIRMWARE_RV32		"airoha/en7581_npu_rv32.bin"
+#define NPU_EN7581_7996_FIRMWARE_DATA	"airoha/en7581_MT7996_npu_data.bin"
+#define NPU_EN7581_7996_FIRMWARE_RV32	"airoha/en7581_MT7996_npu_rv32.bin"
+#define NPU_AN7583_FIRMWARE_DATA		"airoha/an7583_npu_data.bin"
+#define NPU_AN7583_FIRMWARE_RV32		"airoha/an7583_npu_rv32.bin"
+#define NPU_EN7581_FIRMWARE_RV32_MAX_SIZE	0x200000
+#define NPU_EN7581_FIRMWARE_DATA_MAX_SIZE	0x10000
+#define NPU_DUMP_SIZE				512
+
+#define REG_NPU_LOCAL_SRAM		0x0
+
+#define NPU_PC_BASE_ADDR		0x305000
+#define REG_PC_DBG(_n)			(0x305000 + ((_n) * 0x100))
+
+#define NPU_CLUSTER_BASE_ADDR		0x306000
+
+#define REG_CR_BOOT_TRIGGER		(NPU_CLUSTER_BASE_ADDR + 0x000)
+#define REG_CR_BOOT_CONFIG		(NPU_CLUSTER_BASE_ADDR + 0x004)
+#define REG_CR_BOOT_BASE(_n)		(NPU_CLUSTER_BASE_ADDR + 0x020 + ((_n) << 2))
+
+#define NPU_MBOX_BASE_ADDR		0x30c000
+
+#define REG_CR_MBOX_INT_STATUS		(NPU_MBOX_BASE_ADDR + 0x000)
+#define MBOX_INT_STATUS_MASK		BIT(8)
+
+#define REG_CR_MBOX_INT_MASK(_n)	(NPU_MBOX_BASE_ADDR + 0x004 + ((_n) << 2))
+#define REG_CR_MBQ0_CTRL(_n)		(NPU_MBOX_BASE_ADDR + 0x030 + ((_n) << 2))
+#define REG_CR_MBQ8_CTRL(_n)		(NPU_MBOX_BASE_ADDR + 0x0b0 + ((_n) << 2))
+#define REG_CR_NPU_MIB(_n)		(NPU_MBOX_BASE_ADDR + 0x140 + ((_n) << 2))
+
+#define NPU_WLAN_BASE_ADDR		0x30d000
+
+#define REG_IRQ_STATUS			(NPU_WLAN_BASE_ADDR + 0x030)
+#define REG_IRQ_RXDONE(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 2) + 0x034)
+#define NPU_IRQ_RX_MASK(_n)		((_n) == 1 ? BIT(17) : BIT(16))
+
+#define REG_TX_BASE(_n)			(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x080)
+#define REG_TX_DSCP_NUM(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x084)
+#define REG_TX_CPU_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x088)
+#define REG_TX_DMA_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x08c)
+
+#define REG_RX_BASE(_n)			(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x180)
+#define REG_RX_DSCP_NUM(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x184)
+#define REG_RX_CPU_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x188)
+#define REG_RX_DMA_IDX(_n)		(NPU_WLAN_BASE_ADDR + ((_n) << 4) + 0x18c)
+
+#define NPU_TIMER_BASE_ADDR		0x310100
+#define REG_WDT_TIMER_CTRL(_n)		(NPU_TIMER_BASE_ADDR + ((_n) * 0x100))
+#define WDT_EN_MASK			BIT(25)
+#define WDT_INTR_MASK			BIT(21)
+
+#define CR_NPU_MIB31			(NPU_MBOX_BASE_ADDR + 0x1BC)
+#define INIT_COMPELET			0xCCCCCCCC
+#define WLAN_MAX_STATS_SIZE		4408
+
+enum {
+	NPU_OP_SET = 1,
+	NPU_OP_SET_NO_WAIT,
+	NPU_OP_GET,
+	NPU_OP_GET_NO_WAIT,
+};
+
+enum {
+	NPU_FUNC_WIFI,
+	NPU_FUNC_TUNNEL,
+	NPU_FUNC_NOTIFY,
+	NPU_FUNC_DBA,
+	NPU_FUNC_TR471,
+	NPU_FUNC_PPE,
+};
+
+enum {
+	NPU_MBOX_ERROR,
+	NPU_MBOX_SUCCESS,
+};
+
+enum {
+	PPE_FUNC_SET_WAIT,
+	PPE_FUNC_SET_WAIT_HWNAT_INIT,
+	PPE_FUNC_SET_WAIT_HWNAT_DEINIT,
+	PPE_FUNC_SET_WAIT_API,
+	PPE_FUNC_SET_WAIT_FLOW_STATS_SETUP,
+	PPE_FUNC_SET_WAIT_L4S_SETUP,
+};
+
+enum {
+	PPE2_SRAM_SET_ENTRY,
+	PPE_SRAM_SET_ENTRY,
+	PPE_SRAM_SET_VAL,
+	PPE_SRAM_RESET_VAL,
+};
+
+enum {
+	QDMA_WAN_ETHER = 1,
+	QDMA_WAN_PON_XDSL,
+};
+
+struct airoha_npu_fw {
+	const char *name;
+	int max_size;
+};
+
+struct airoha_npu_soc_data {
+	struct airoha_npu_fw fw_rv32;
+	struct airoha_npu_fw fw_data;
+};
+
+#define MBOX_MSG_FUNC_ID	GENMASK(14, 11)
+#define MBOX_MSG_STATIC_BUF	BIT(5)
+#define MBOX_MSG_STATUS		GENMASK(4, 2)
+#define MBOX_MSG_DONE		BIT(1)
+#define MBOX_MSG_WAIT_RSP	BIT(0)
+
+#define PPE_TYPE_L2B_IPV4	2
+#define PPE_TYPE_L2B_IPV4_IPV6	3
+
+struct ppe_mbox_data {
+	u32 func_type;
+	u32 func_id;
+	union {
+		struct {
+			u8 cds;
+			u8 xpon_hal_api;
+			u8 wan_xsi;
+			u8 ct_joyme4;
+			u8 max_packet;
+			u8 rsv[3];
+			u32 ppe_type;
+			u32 wan_mode;
+			u32 wan_sel;
+		} init_info;
+		struct {
+			u32 func_id;
+			u32 size;
+			u32 data;
+		} set_info;
+		struct {
+			u32 npu_stats_addr;
+			u32 foe_stats_addr;
+		} stats_info;
+		struct {
+			u32 set_cmd;
+			u32 set_water_mark;
+		} l4s_info;
+	};
+};
+struct tr471_mbox_data {
+	int func_type;
+	char private[60];
+};
+
+struct wlan_mbox_data {
+	u32 ifindex:4;
+	u32 func_type:4;
+	u32 func_id;
+	DECLARE_FLEX_ARRAY(u8, d);
+	union {
+		u32 data;
+		struct {
+			u32 dir;
+			u32 in_counter_addr;
+			u32 out_status_addr;
+			u32 out_counter_addr;
+		} txrx_addr;
+		u8 stats[WLAN_MAX_STATS_SIZE];
+	};
+};
+int airoha_npu_send_msg(struct airoha_npu *npu, int func_id, void *p, int size);
+int airoha_npu_l4s_setup(struct airoha_npu *npu, u32 set_cmd, u32 set_water_mark);
+int airoha_npu_wlan_msg_get(struct airoha_npu *npu, int ifindex, enum airoha_npu_wlan_get_cmd func_id, void *data, int data_len, gfp_t gfp);
+int airoha_npu_wlan_msg_send(struct airoha_npu *npu, int ifindex,
+				    enum airoha_npu_wlan_set_cmd func_id,
+				    void *data, int data_len, gfp_t gfp);
+
+int airoha_npu_extra_init(struct airoha_npu *npu);
+
+
+#endif /* AIROHA_NPU_H */
