--- a/sound/soc/mediatek/Kconfig
+++ b/sound/soc/mediatek/Kconfig
@@ -27,7 +27,7 @@ config SND_SOC_MT2701_CS42448
 config SND_SOC_MT2701_WM8960
 	tristate "ASoc Audio driver for MT2701 with WM8960 codec"
 	depends on SND_SOC_MT2701 && I2C
-	select SND_SOC_WM8960
+	select SND_SOC_SLIC_DUMMY
 	help
 	  This adds ASoC driver for Mediatek MT2701 boards
 	  with the WM8960 codecs.
--- a/sound/soc/mediatek/mt2701/mt2701-wm8960.c
+++ b/sound/soc/mediatek/mt2701/mt2701-wm8960.c
@@ -56,7 +56,7 @@ SND_SOC_DAILINK_DEFS(capture,
 
 SND_SOC_DAILINK_DEFS(codec,
 	DAILINK_COMP_ARRAY(COMP_CPU("I2S0")),
-	DAILINK_COMP_ARRAY(COMP_CODEC(NULL, "wm8960-hifi")),
+	DAILINK_COMP_ARRAY(COMP_CODEC(NULL, "slic-dummy-aif")),
 	DAILINK_COMP_ARRAY(COMP_EMPTY()));
 
 static struct snd_soc_dai_link mt2701_wm8960_dai_links[] = {
@@ -138,19 +138,22 @@ static int mt2701_wm8960_machine_probe(s
 		dai_link->codecs->of_node = codec_node;
 	}
 
+#if 0 // D2 Mod
 	ret = snd_soc_of_parse_audio_routing(card, "audio-routing");
 	if (ret) {
 		dev_err(&pdev->dev, "failed to parse audio-routing: %d\n", ret);
 		goto put_codec_node;
 	}
-
+#endif
 	ret = devm_snd_soc_register_card(&pdev->dev, card);
 	if (ret)
 		dev_err(&pdev->dev, "%s snd_soc_register_card fail %d\n",
 			__func__, ret);
 
+#if 0 // D2 Mod
 put_codec_node:
 	of_node_put(codec_node);
+#endif
 put_platform_node:
 	of_node_put(platform_node);
 	return ret;
