commit e6155b2d75cca051466fb8f9a496581f9af1c205
Author: Daniel Golle <daniel@makrotopia.org>
Date:   Thu Mar 20 21:44:17 2025 +0000

    hynitron-cst816x: wip

--- a/drivers/input/touchscreen/hynitron-cst816x.c
+++ b/drivers/input/touchscreen/hynitron-cst816x.c
@@ -10,6 +10,7 @@
 #include <linux/gpio/consumer.h>
 #include <linux/i2c.h>
 #include <linux/input.h>
+#include <linux/input/touchscreen.h>
 #include <linux/interrupt.h>
 #include <linux/module.h>
 
@@ -32,6 +33,8 @@ struct cst816x_priv {
 	struct gpio_desc *reset;
 	struct input_dev *input;
 	struct cst816x_event_map event_map[16];
+	unsigned long active_events;
+	struct touchscreen_properties prop;
 };
 
 static int cst816x_parse_dt_event_map(struct device *dev,
@@ -123,6 +126,8 @@ static int cst816x_register_input(struct
 
 	input_set_abs_params(priv->input, ABS_X, 0, 240, 0, 0);
 	input_set_abs_params(priv->input, ABS_Y, 0, 240, 0, 0);
+	input_set_capability(priv->input, EV_KEY, BTN_TOUCH);
+	touchscreen_parse_properties(priv->input, false, &priv->prop);
 
 	return input_register_device(priv->input);
 }
@@ -139,23 +144,28 @@ static irqreturn_t cst816x_irq_cb(int ir
 {
 	struct cst816x_priv *priv = cookie;
 	struct cst816x_touch_info info;
+	int i;
 
 	if (!cst816x_process_touch(priv, &info))
 		return IRQ_HANDLED;
 
-	if (info.touch) {
-		input_report_abs(priv->input, ABS_X, info.abs_x);
-		input_report_abs(priv->input, ABS_Y, info.abs_y);
-		input_report_key(priv->input, BTN_TOUCH, 1);
-	}
+	touchscreen_report_pos(priv->input, &priv->prop,
+			       info.abs_x, info.abs_y, false);
+
+	input_report_key(priv->input, BTN_TOUCH, !!info.touch);
 
 	if (info.gesture) {
 		input_report_key(priv->input,
 				 priv->event_map[info.gesture & 0x0F].code,
-				 info.touch);
+				 1);
 
-		if (!info.touch)
-			input_report_key(priv->input, BTN_TOUCH, 0);
+		priv->active_events |= BIT(info.gesture & 0xF);
+	} else {
+		for_each_set_bit(i, &priv->active_events, 0xF)
+			input_report_key(priv->input,
+					 priv->event_map[i].code,
+					 0);
+		priv->active_events = 0;
 	}
 
 	input_sync(priv->input);
